config/helpers/bootstrap.go

```diff
@@ -218,65 +220,69 @@ func createResourceFromFS(ctx context.Context, client dynamic.Interface, mapper
 			}
 
 			if _, exists := existing.GetAnnotations()[annotationCreateOnlyKey]; exists {
-				klog.Infof(
-					"Skipping update of %s %s because it has the create-only annotation",
-					gvk,
-					tenancyhelper.QualifiedObjectName(existing),
-				)
-
+				logger.Info("Skipping update of object because it has the create-only annotation", "object", tenancyhelper.QualifiedObjectName(existing), "gvk", gvk)
 				return nil
 			}
 
 			u.SetResourceVersion(existing.GetResourceVersion())
 			if _, err = client.Resource(m.Resource).Namespace(u.GetNamespace()).Update(ctx, u, metav1.UpdateOptions{}); err != nil {
 				return fmt.Errorf("could not update %s %s: %w", gvk.Kind, tenancyhelper.QualifiedObjectName(existing), err)
 			} else {
-				klog.Infof("Updated %s %s", gvk, tenancyhelper.QualifiedObjectName(existing))
+				logger.Info("Updated object", "object", tenancyhelper.QualifiedObjectName(existing), "gvk", gvk)
 				return nil
 			}
 		}
 		return err
 	}
 
-	klog.Infof("Bootstrapped %s %s", gvk.Kind, tenancyhelper.QualifiedObjectName(upserted))
-
+	logger.Info("Bootstrapped object", "object", tenancyhelper.QualifiedObjectName(upserted), "gvk", gvk)
 	return nil
 }
 
 func BindRootAPIs(ctx context.Context, kcpClient kcpclient.Interface, exportNames ...string) error {
+	logger := klog.FromContext(ctx)
 	for _, exportName := range exportNames {
-		binding := &apisv1alpha1.APIBinding{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: exportName,
-			},
-			Spec: apisv1alpha1.APIBindingSpec{
-				Reference: apisv1alpha1.ExportReference{
-					Workspace: &apisv1alpha1.WorkspaceExportReference{
-						Path:       tenancyv1alpha1.RootCluster.String(),
-						ExportName: exportName,
+		if err := wait.PollImmediateInfiniteWithContext(ctx, time.Second, func(ctx context.Context) (bool, error) {
+			binding := &apisv1alpha1.APIBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: exportName,
+				},
+				Spec: apisv1alpha1.APIBindingSpec{
+					Reference: apisv1alpha1.ExportReference{
+						Workspace: &apisv1alpha1.WorkspaceExportReference{
+							Path:       tenancyv1alpha1.RootCluster.String(),
+							ExportName: exportName,
+						},
 					},
 				},
-			},
-		}
+			}
 
-		created, err := kcpClient.ApisV1alpha1().APIBindings().Create(ctx, binding, metav1.CreateOptions{})
-		if err == nil {
-			klog.V(2).Infof("Created API binding %s|%s", logicalcluster.From(created), created.Name)
-			continue
-		}
-		if !apierrors.IsAlreadyExists(err) {
-			return err
-		}
+			created, err := kcpClient.ApisV1alpha1().APIBindings().Create(ctx, binding, metav1.CreateOptions{})
+			if err == nil {
+				logger.V(2).Info("Created APIBinding", "APIBinding", fmt.Sprintf("%s|%s", logicalcluster.From(created), created.Name))
+				return true, nil
+			}
+			if !apierrors.IsAlreadyExists(err) {
```

Comment:
What errors were you seeing here that required the poll?

Reply:
This is pre-readiness I would assume. So we need a poll. Exports are not ready potentially (expectedly).